### PR TITLE
Introduced foreman-debug subpackage

### PR DESCRIPTION
**Pull request by lzap**
Replaces https://github.com/theforeman/foreman-packaging/pull/541

### CODE REVIEW

File: debian/precise/foreman/foreman-debug.dirs [hunk 1]
```diff
@@ -0,0 +1 @@
+usr/share/foreman/script/foreman-debug.d/
```
Comment (GregSutcliffe) [inline debian/precise/foreman/foreman-debug.dirs]: (Just to keep the discussion from the old PR) What's this for? :)

Comment (lzap) [inline debian/precise/foreman/foreman-debug.dirs]: It's an extension point, example plugin: https://github.com/Katello/katello/blob/master/deploy/script/katello-debug.sh

Comment (domcleal) [inline debian/precise/foreman/foreman-debug.dirs]: Or https://github.com/theforeman/smart-proxy/blob/develop/extra/foreman-debug-proxy

**Comment by lzap** [comment]
Build done, where do I get the deb files? http://ci.theforeman.org/view/Packaging/job/packaging_build_deb_coreproject/1647/

**Comment by domcleal** [comment]
Don't forget to add the foreman{,-proxy} deps on foreman-debug.

**Comment by domcleal** [comment]
(Maybe Recommends?)

**Comment by domcleal** [comment]
And ensure that foreman-proxy's debug.d file is placed in ~foreman's debug.d dir.

File: debian/precise/foreman/foreman-debug.dirs (hunk 1 above)
Comment (GregSutcliffe) [inline debian/precise/foreman/foreman-debug.dirs]: This should be part of the main package if the proxy is going to drop files into it.

**Comment by lzap** [comment]
Adding hard dependency, that's better (users will ask where `foreman-debug` actually is otherwise).

Will amend the foreman-proxy change in this patch as well.

**Comment by domcleal** [comment]
Recommends will be installed by default though, but allows users to remove the package to create a minimal installation.

File: debian/precise/foreman/foreman-debug.dirs (hunk 1 above)
Comment (domcleal) [inline debian/precise/foreman/foreman-debug.dirs]: In the RPMs, foreman-proxy depends on foreman-debug so the package would exist.  Even if it's a recommends and foreman-debug's not installed, will the proxy/plugin package install the file OK?  (It'd just be unused.)

**Comment by lzap** [comment]
Oh okay, will change that.

**Comment by lzap** [comment]
Amended incl. the proxy change (man that was close to reformatting). Triggering the build.

**Comment by lzap** [comment]
http://ci.theforeman.org/view/Packaging/job/packaging_build_deb_coreproject/1648/

**Comment by domcleal** [comment]
And add foreman-debug to foreman-proxy's recommends?

**Comment by lzap** [comment]
*facepalm*

**Comment by lzap** [comment]
http://ci.theforeman.org/view/Packaging/job/packaging_build_deb_coreproject/1649/

**Comment by domcleal** [comment]
Looks good.  We should test the installation without foreman-debug just to confirm the inline discussion above.

**Comment by lzap** [comment]
Problem A: `apt-get upgrade` did not install recommended package for me.

Problem B: `foreman-debug` package does not contain any files :-)

```
root@fwheezy:~# dpkg -L foreman-debug | grep debug
/usr/share/doc/foreman-debug
/usr/share/doc/foreman-debug/changelog.Debian.gz
/usr/share/doc/foreman-debug/copyright
/usr/share/doc/foreman-debug/changelog.gz
```

**Comment by lzap** [comment]
I left the grep in the pipeline but there is nothing else left.

I don't belive I've installed correct build.

**Comment by lzap** [comment]
Hmmm it really looks like the version is correct. http://ci.theforeman.org/job/packaging_build_deb_coreproject/label=debian,os=wheezy/1650/console

    Setting up foreman-debug (9999-wheezy+scratchbuild+201502131319) ...

Help :-)

**Comment by domcleal** [comment]
foreman-debug.{dirs,install} appear to be outside of the 'foreman' subdirectory for trusty/wheezy, but appear to be in the correct place for precise.

File: debian/precise/foreman/foreman-debug.dirs (hunk 1 above)
Comment (GregSutcliffe) [inline debian/precise/foreman/foreman-debug.dirs]: I'm not 100% sure that an `install` entry without a `dirs` entry will work as described, but it's fairly likely. we can test it :)

**Comment by GregSutcliffe** [comment]
Agreed, your dirs are wrong. We'll also need scratchbuilds of foreman-proxy to check the debug dir is written correctly - I only see foreman scratchbuilds at the  moment.

**Comment by lzap** [comment]
http://ci.theforeman.org/view/Packaging/job/packaging_build_deb_coreproject/1660/
http://ci.theforeman.org/view/Packaging/job/packaging_build_deb_coreproject/1661/

**Comment by domcleal** [comment]
http://ci.theforeman.org/view/Packaging/job/packaging_build_deb_coreproject/1664/ is the right build for the proxy, replaces the second link.

**Comment by lzap** [comment]
I think we are fine, it works. Uploaded a tarball actually to investigate if all things got collected. One issue tho (out of scope for this?) - the foreman-debug is installed twice - sbin and then script directories. We should either do a symlink or remove the latter.

```
root@fwheezy:~# dpkg -L foreman | grep debug
/usr/share/foreman/script/foreman-debug
/usr/share/foreman/lib/foreman/gettext/debug.rb
root@fwheezy:~# dpkg -L foreman-debug | grep debug
/usr/share/doc/foreman-debug
/usr/share/doc/foreman-debug/changelog.Debian.gz
/usr/share/doc/foreman-debug/copyright
/usr/share/doc/foreman-debug/changelog.gz
/usr/share/foreman/script/foreman-debug.d
/usr/sbin/foreman-debug
root@fwheezy:~# dpkg -L foreman-proxy | grep debug
/usr/share/foreman/script/foreman-debug.d
/usr/share/foreman/script/foreman-debug.d/foreman-debug-proxy
root@fwheezy:~# ls /usr/sbin/foreman-debug -la
-rwxr-xr-x 1 root root 9103 Feb 17 04:54 /usr/sbin/foreman-debug
root@fwheezy:~# ls /usr/share/foreman/script/foreman-debug -la
-rwxr-xr-x 1 foreman foreman 9103 Feb 17 04:54 /usr/share/foreman/script/foreman-debug
root@fwheezy:~# diff /usr/sbin/foreman-debug /usr/share/foreman/script/foreman-debug
root@fwheezy:~# foreman-debug


 HOSTNAME: fwheezy.zzz.lan
       OS: debian
  RELEASE: 7.8
  FOREMAN: 1.8.0-develop
     RUBY: ruby 1.9.3p194 (2012-04-20 revision 35410) [x86_64-linux]
   PUPPET: 2.7.23


A debug file has been created: /tmp/foreman-debug-QaA48.tar.xz (301484 bytes)

You may want to upload the tarball to our public server via rsync. There is a
write only directory (readable only by Foreman core developers) for that. Note
the rsync transmission is UNENCRYPTED:

  rsync /tmp/foreman-debug-QaA48.tar.xz rsync://theforeman.org/debug-incoming

Do you want to do this now? [y/N] y
Uploading...

The tarball has been uploaded, please contact us on mailing list or IRC.
```

**Comment by domcleal** [comment]
@lzap all tested & works great, except for the proxy part.  Looks like the RPMs have an identical problem I missed when reviewing that.

https://github.com/theforeman/foreman/blob/develop/script/foreman-debug#L265 checks for *.sh extension files, but names in both the deb and RPM don't have a .sh extension.

**Comment by lzap** [comment]
Oh I want to fix that in the script itself and include all the executable files and fix this in foreman-proxy git repo (the file currently does not have proper permissions).

**Comment by lzap** [comment]
Proxy fix: https://github.com/theforeman/smart-proxy/pull/259

**Comment by lzap** [comment]
Foreman core fix: https://github.com/theforeman/foreman/pull/2178

File: debian/precise/foreman/foreman-debug.dirs (hunk 1 above)
Comment (domcleal) [inline debian/precise/foreman/foreman-debug.dirs]: Tests fine for me.

**Comment by domcleal** [comment]
Great, thanks for implementing this @lzap!  One comment remains on the core fix, but altogether this works fine.

**Pull request merged**